rbi/annotations/kredis.rbi

```diff
@@ -0,0 +1,183 @@
+  def scalar(key, typed: nil, default: nil, config: nil, after_change: nil, expires_in: nil); end
+
+  sig { params(key: T.untyped, typed: T.untyped, config: T.untyped, after_change: T.untyped,).returns(Kredis::Types::Set[T.untyped]) }
+  def set(key, typed: nil, config: nil, after_change: nil); end
+
+  sig { params(key: T.untyped, config: T.untyped, after_change: T.untyped).returns(Kredis::Types::Slots) }
+  def slot(key, config: nil, after_change: nil); end
+
+  sig { params(key: T.untyped, available: T.untyped, config: T.untyped, after_change: T.untyped).returns(Kredis::Types::Slots) }
+  def slots(key, available:, config: nil, after_change: nil); end
+
+  sig { params(key: T.untyped, default: T.untyped, config: T.untyped, after_change: T.untyped, expires_in: T.untyped).returns(Kredis::Types::Scalar[T.nilable(::String)]) }
+  def string(key, default: nil, config: nil, after_change: nil, expires_in: nil); end
+
+  sig { params(key: T.untyped, typed: T.untyped, limit: T.untyped, config: T.untyped, after_change: T.untyped).returns(Kredis::Types::UniqueList[T.untyped]) }
+  def unique_list(key, typed: nil, limit: nil, config: nil, after_change: nil); end
+end
+
+class Kredis::Types::Counter
+  sig { params(by: Integer).returns(Integer) }
+  def increment(by: 1); end
+
+  sig { params(by: Integer).returns(Integer) }
+  def decrement(by: 1); end
+
+  sig { returns(Integer) }
+  def value; end
+end
+
+class Kredis::Types::Cycle
+  extend T::Generic
+
+  ValueType = type_member
+
+  sig { returns(ValueType) }
+  def value; end
+end
+
+class Kredis::Types::Enum
+  extend T::Generic
+
+  ValueType = type_member
+
+  sig { returns(ValueType) }
+  def value; end
+
+  sig { returns(T::Array[ValueType]) }
+  def values; end
+end
+
+class Kredis::Types::Flag
+  sig { returns(T::Boolean) }
+  def marked?; end
+end
+
+class Kredis::Types::Hash
+  extend T::Generic
+
+  ValueType = type_member
+
+  sig { params(key: T.any(String, Symbol)).returns(ValueType) }
+  def [](key); end
+
+  sig { returns(ActiveSupport::HashWithIndifferentAccess) }
+  def entries; end
+
+  sig { returns(T::Array[String]) }
+  def keys; end
+
+  sig { returns(T::Array[ValueType]) }
+  def values; end
+
+  sig { params(keys: T.any(String, Symbol, T::Array[T.any(String, Symbol)])).returns(ValueType) }
+  def values_at(*keys); end
+end
+
+class Kredis::Types::List
+  extend T::Generic
+
+  ValueType = type_member
+
+  sig { returns(T::Array[ValueType]) }
+  def elements; end
+end
+
+class Kredis::Types::Proxy
+  # @method_missing: subclasses delegate (via `proxying` class method) to Proxy, https://github.com/rails/kredis/blob/v1.3.0/lib/kredis/types/proxy.rb#L23
+  sig { params(_arg0: T.untyped, _arg1: T.untyped, _arg2: T.nilable(T.proc.void)).returns(T::Boolean) }
+  def exists?(*_arg0, **_arg1, &_arg2); end
+end
+
+class Kredis::Types::Scalar
+  extend T::Generic
```

Comment:
The runtime check doesn't like these declarations, but not sure how to fix. Seems useful to know what values we'll be getting back, this is what happens at runtime:
```ruby
i = Kredis.integer "ex1"
#<Kredis::Types::Scalar:0x0000000109504010                   
 @default=nil,                                               
 @expires_in=nil,                                            
 @key="ex1",                                                 
 @proxy=#<Kredis::Types::Proxy:0x000000010956ff90 @key="ex1", @redis=#<Redis client v5.0.5 for redis://127.0.0.1:6379/0>>,
 @typed=:integer>  
i.value = "42"
# Kredis Proxy (1.4ms)  SET ex1 [42] 
=> "42"                                            
i.value
# Kredis Proxy (1.1ms)  GET ex1  
=> 42
```

Reply:
I don't think we can go this route. Generic is a runtime behaviour that can't be only represented from RBIs.

Having an annotation like this would make the static checker not error for usages such as `Kredis[Integer].new(42)` which is an error because Kredis doesn't actually extend `T::Generic` and won't have access to [#[]](https://github.com/sorbet/sorbet/blob/052762b0658d99d0c622d535b181b93ff7af9267/gems/sorbet-runtime/lib/types/generic.rb#L11).

I think only alternative without changing the library is keeping the signatures as `T.untyped`. Which may mean these annotations aren't necessary. 

cc @Morriar if you have anything to add. I was slightly confused because of https://github.com/Shopify/rbi-central/pull/64

Reply:
Bummer. In that case I can cut anything using generics, some signatures are still useful here IMO (eg the factory methods for `Kredis::Types`). Will update.